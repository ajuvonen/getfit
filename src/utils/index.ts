import {unref} from 'vue';
import {helpers} from '@vuelidate/validators';
import type {ErrorObject} from '@vuelidate/core';
import {v4 as uuid} from 'uuid';
import {clone} from 'remeda';
import {Intensity, type ScheduleSettings, type Training} from '@/types';
import {ACTIVITIES, COLORS} from '@/constants';

export const getIcon = (activityName: string) => `$${activityName}`;

export const getIntensityColor = (intensity: Intensity, alpha: number = 1) => {
  const hexAlpha = alpha < 1 ? Math.round(alpha * 255).toString(16) : '';
  switch (intensity) {
    case Intensity.NORMAL:
      return COLORS.intensityNormal + hexAlpha;
    case Intensity.DEMANDING:
      return COLORS.intensityDemanding + hexAlpha;
    case Intensity.HEAVY:
      return COLORS.intensityHeavy + hexAlpha;
    default:
      return COLORS.intensityLight + hexAlpha;
  }
};

export const getEmptySettings = (
  initialSettings: Partial<ScheduleSettings> = {},
): ScheduleSettings => ({
  name: '',
  startsOnSunday: false,
  startDate: null,
  actualWeekNumbering: false,
  availableActivities: clone(ACTIVITIES),
  defaultStartTime: {
    hours: 12,
    minutes: 0,
    seconds: 0,
  },
  defaultDuration: 1,
  defaultUnitOfDuration: 'h',
  darkMode: 'auto',
  decoratedCards: true,
  ...initialSettings,
});

export const getEmptyTraining = (initialTraining: Partial<Training> = {}): Training => ({
  id: uuid(),
  weekId: uuid(),
  activity: '',
  dayIndex: 0,
  title: '',
  instructions: '',
  duration: 1,
  unitOfDuration: 'h',
  intensity: Intensity.NORMAL,
  location: '',
  completed: false,
  rating: null,
  ...initialTraining,
});

export const decimalRegex = helpers.regex(/^\d+(.(00?|25|50?|75))?$/);

export const getValidationErrors = (errors: ErrorObject[]) =>
  errors.map((error) => unref(error.$message));

export const isDurationTime = (value: string) => ['h', 'm'].includes(value);
