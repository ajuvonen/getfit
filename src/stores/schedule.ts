import {defineStore} from 'pinia';
import {useStorage} from '@vueuse/core';
import {v4 as uuidv4} from 'uuid';
import {clone, lensProp, over} from 'ramda';
import {DateTime} from 'luxon';
import type {ScheduleSettings, Training, Week} from '@/types';
import {ACTIVITIES} from '@/constants';
import {roundNearestQuarter} from '@/utils';
import {computed, watch} from 'vue';

const getEmptySchedule = (): ScheduleSettings => ({
  name: '',
  startsOnSunday: false,
  startDate: null,
  actualWeekNumbering: false,
  availableActivities: ACTIVITIES.map(({value}) => value),
  defaultStartTime: {
    hours: 12,
    minutes: 0,
    seconds: 0,
  },
  defaultDuration: 1,
  unitOfTime: 'h',
});

export const useScheduleStore = defineStore('schedule', () => {
  // State refs
  const settings = useStorage('getfit-settings', getEmptySchedule(), localStorage, {
    mergeDefaults: true,
  });

  const weeks = useStorage('getfit-schedule', [] as Week[], localStorage, {mergeDefaults: true});

  // Computed getters
  const getTargetWeekAndTraining = computed(() => (weekId: string, trainingId?: string) => {
    const targetWeek = weeks.value.find(({id}) => id === weekId) as Week;
    const targetTraining = trainingId
      ? targetWeek.trainings.find(({id}) => id === trainingId)
      : undefined;
    return [targetWeek, targetTraining] as [Week, Training | undefined];
  });

  // Actions
  const addWeek = () => {
    weeks.value.push({
      id: uuidv4(),
      trainings: [],
    });
  };

  const deleteWeek = (weekId: string) => {
    weeks.value = weeks.value.filter(({id}) => id !== weekId);
  };

  const copyWeek = (weekId: string) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    const newWeekId = uuidv4();
    const cloneWeek: Week = {
      id: newWeekId,
      trainings: targetWeek.trainings.map((training) => ({
        ...training,
        id: uuidv4(),
        weekId: newWeekId,
      })),
    };
    weeks.value.push(cloneWeek);
  };

  const addOrEditTraining = (training: Training) => {
    const [targetWeek] = getTargetWeekAndTraining.value(training.weekId);
    const targetIndex = targetWeek.trainings.findIndex(({id}) => id === training.id);
    if (targetIndex >= 0) {
      targetWeek.trainings[targetIndex] = training;
    } else {
      targetWeek.trainings.push(training);
    }
  };

  const deleteTraining = (training: Training) => {
    const [targetWeek] = getTargetWeekAndTraining.value(training.weekId);
    targetWeek.trainings = targetWeek.trainings.filter(({id}) => id !== training.id);
  };

  const moveTraining = (training: Training, newWeekId: string, dayIndex: number) => {
    const [originalWeek, targetTraining] = getTargetWeekAndTraining.value(
      training.weekId,
      training.id,
    ) as [Week, Training];
    const [targetWeek] = getTargetWeekAndTraining.value(newWeekId);
    originalWeek.trainings = originalWeek.trainings.filter(({id}) => id !== targetTraining.id);
    targetTraining.weekId = newWeekId;
    targetTraining.dayIndex = dayIndex;
    targetWeek.trainings.push(targetTraining);
  };

  const reorderTrainings = (weekId: string, trainings: Training[]) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    targetWeek.trainings = trainings;
  };

  const copyTraining = (training: Training, weekId: string, dayIndex: number) => {
    const [targetWeek] = getTargetWeekAndTraining.value(weekId);
    const clonedTraining: Training = {
      ...clone(training),
      id: uuidv4(),
      weekId,
      dayIndex,
    };
    targetWeek.trainings.push(clonedTraining);
  };

  // Watchers
  watch(
    () => settings.value.unitOfTime,
    (newValue, oldValue) => {
      if (newValue && newValue !== oldValue) {
        const multiplier = newValue === 'h' ? 1 / 60 : 60;
        const precision = newValue === 'h' ? 2 : 0;
        settings.value.defaultDuration = roundNearestQuarter(
          settings.value.defaultDuration * multiplier,
          precision,
        );
        weeks.value = weeks.value.map((week) => ({
          ...week,
          trainings: week.trainings.map(
            over(lensProp('duration'), (duration) =>
              roundNearestQuarter(duration * multiplier, precision),
            ),
          ),
        }));
      }
    },
  );

  watch(
    () => settings.value.startsOnSunday,
    (newValue) => {
      const parsedValue = newValue || false;
      if (settings.value.startDate) {
        const date = DateTime.fromJSDate(settings.value.startDate);
        if (parsedValue) {
          settings.value.startDate = date.minus({days: 1}).toJSDate();
        } else {
          settings.value.startDate = date.plus({days: 1}).toJSDate();
        }
      }
    },
  );

  // Reset
  const $reset = () => {
    settings.value = getEmptySchedule();
    weeks.value = [];
  };

  return {
    settings,
    weeks,
    getTargetWeekAndTraining,
    addWeek,
    deleteWeek,
    copyWeek,
    addOrEditTraining,
    deleteTraining,
    moveTraining,
    reorderTrainings,
    copyTraining,
    $reset,
  };
});
