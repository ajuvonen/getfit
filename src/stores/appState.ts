import {defineStore} from 'pinia';
import {v4 as uuidv4} from 'uuid';
import type {Training} from '@/types';
import {clone} from 'ramda';
import {useScheduleStore} from './schedule';

export const useAppStateStore = defineStore('appState', {
  state: () => ({
    summaryShown: [] as string[],
    trainingDialogOpen: false,
    trainingData: {} as Training,
  }),
  getters: {
    isSummaryShown: (state) => (trainingId: string) => state.summaryShown.includes(trainingId),
  },
  actions: {
    openNewTrainingDialog(weekId: string, dayIndex: number) {
      this.trainingData = {
        id: uuidv4(),
        weekId,
        dayIndex,
        duration: 0,
        activity: '',
        description: '',
        title: '',
        intensity: 0,
        completionSummary: '',
        location: '',
      };
      this.trainingDialogOpen = true;
    },
    openEditTrainingDialog(training: Training) {
      const scheduleStore = useScheduleStore();
      const clonedTraining = clone(training);
      if (
        clonedTraining.activity &&
        !scheduleStore.schedule.availableActivities.includes(clonedTraining.activity)
      ) {
        clonedTraining.activity = '';
      }
      this.trainingData = clonedTraining;
      this.trainingDialogOpen = true;
    },
    closeEditTrainingDialog() {
      this.trainingDialogOpen = false;
    },
    toggleSummaryShown(trainingId: string) {
      if (this.isSummaryShown(trainingId)) {
        this.summaryShown = this.summaryShown.filter((id) => id !== trainingId);
      } else {
        this.summaryShown.push(trainingId);
      }
    },
  },
});
