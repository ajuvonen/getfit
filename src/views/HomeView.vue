<script setup lang="ts">
import {computed, ref} from 'vue';
import {storeToRefs} from 'pinia';
import {PHOTO_CREDITS} from '@/constants';
import {useScheduleStore} from '@/stores/schedule';
import InfoDialog from '@/components/InfoDialog.vue';
import DailyAgenda from '@/components/DailyAgenda.vue';

const scheduleStore = useScheduleStore();
const {weeks, settings} = storeToRefs(scheduleStore);

const creditDialogOpen = ref(false);

const agendaAvailable = computed(() => weeks.value.length && settings.value.startDate);
</script>

<template>
  <div class="home__overlay text-body-1 text-grey-lighten-5">
    <h1 class="text-h3 text-center">{{ $t(agendaAvailable ? 'home.dailyAgenda' : 'home.title') }}</h1>
    <DailyAgenda v-if="agendaAvailable" />
    <template v-else>
      <p class="mt-10">
        {{ $t('home.p1') }}
      </p>
      <p class="mt-4">
        {{ $t('home.p2') }}
      </p>
      <p class="mt-4">
        {{ $t('home.p3') }}
      </p>
    </template>
    <p class="mt-10">
      <v-icon icon="mdi-github" />
      <a class="text-grey-lighten-5 ml-2" href="https://www.github.com/ajuvonen/getfit">GitHub</a>
    </p>
    <v-btn
      variant="text"
      size="x-small"
      class="home__photo-credit"
      @click="creditDialogOpen = true"
    >
      {{ $t('home.credits') }}
    </v-btn>
    <InfoDialog
      :show="creditDialogOpen"
      :title="$t('home.credits')"
      @close="creditDialogOpen = false"
    >
      <template #content>
        <ul>
          <li v-for="{name, link} in PHOTO_CREDITS" :key="name">
            <a :href="link" noreferrer noopener>{{ name }}</a>
          </li>
        </ul>
      </template>
    </InfoDialog>
  </div>
</template>
<style lang="scss" scoped>
.home__overlay {
  position: absolute;
  inset: 0;
  text-align: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 1rem;
}

.home__photo-credit {
  position: absolute;
  bottom: 1rem;
  left: 50%;
  transform: translateX(-50%);
}

ul {
  list-style-type: none;

  a {
    color: inherit;
  }
}
</style>
