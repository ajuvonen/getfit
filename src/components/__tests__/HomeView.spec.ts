import {it, expect, describe, beforeEach} from 'vitest';
import {mount} from '@vue/test-utils';
import {DateTime} from 'luxon';
import {v4 as uuid} from 'uuid';
import {useScheduleStore} from '@/stores/schedule';
import {getEmptyTraining} from '@/utils';
import HomeView from '@/views/HomeView.vue';

describe('HomeView', () => {
  let scheduleStore: ReturnType<typeof useScheduleStore>;

  beforeEach(() => {
    scheduleStore = useScheduleStore();
  });

  it('mounts', () => {
    const wrapper = mount(HomeView);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('mounts with agenda', () => {
    const weekId = uuid();
    scheduleStore.weeks.push({
      id: weekId,
      trainings: [
        getEmptyTraining({
          weekId,
          activity: 'running',
          dayIndex: DateTime.now().weekday - 1,
        }),
      ],
    });
    scheduleStore.settings.startDate = DateTime.now().startOf('week').toJSDate();
    const wrapper = mount(HomeView);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
