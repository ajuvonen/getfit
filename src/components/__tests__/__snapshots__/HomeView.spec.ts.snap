// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeView > mounts 1`] = `
"<div data-v-b4e148ca="" class="home__overlay text-body-1 text-grey-lighten-5">
  <h1 data-v-b4e148ca="" class="text-h3 text-center">Let's GetFit </h1>
  <p data-v-b4e148ca="" class="mt-10">GetFit is a free exercise program planner that allows you to plan a schedule spanning multiple weeks. To get started, navigate to settings and fill some basic info. Then, go to your schedule and start planning activities. If you give your schedule a starting date, a daily agenda will appear on this screen.</p>
  <p data-v-b4e148ca="" class="mt-4">Schedules created with GetFit can be exported to .ics format, which is compatible with most email clients and calendars. You can install GetFit as a web app on your device, in which case you can work also offline. All data persists on your device, and is not sent anywhere.</p>
  <p data-v-b4e148ca="" class="mt-10"><i data-v-b4e148ca="" class="mdi-github mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i><a data-v-b4e148ca="" class="text-grey-lighten-5 ml-2" target="_blank" noopener="" noreferrer="" href="https://www.github.com/ajuvonen/getfit">GitHub</a></p><button data-v-b4e148ca="" type="button" class="v-btn v-theme--light v-btn--density-default v-btn--size-x-small v-btn--variant-text home__photo-credit"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator="">Photo credits</span>
    <!---->
    <!---->
  </button>
  <!---->
  <!---->
</div>"
`;

exports[`HomeView > mounts with agenda 1`] = `
"<div data-v-b4e148ca="" class="home__overlay text-body-1 text-grey-lighten-5">
  <div data-v-c4ee72cd="" class="v-progress-circular v-theme--light mx-auto" style="width: 150px; height: 150px;" role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="NaN"><svg style="transform: rotate(calc(-90deg + 0deg));" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 44.44444444444444 44.44444444444444">
      <circle class="v-progress-circular__underlay" fill="transparent" cx="50%" cy="50%" r="20" stroke-width="4.444444444444445" stroke-dasharray="125.66370614359172" stroke-dashoffset="0"></circle>
      <circle class="v-progress-circular__overlay" fill="transparent" cx="50%" cy="50%" r="20" stroke-width="4.444444444444445" stroke-dasharray="125.66370614359172" stroke-dashoffset="NaN"></circle>
    </svg>
    <div class="v-progress-circular__content">
      <transition-stub data-v-c4ee72cd="" name="fab-transition" mode="out-in" appear="false" persisted="false" css="true"><i data-v-c4ee72cd="" class="mdi-trophy-outline mdi v-icon notranslate v-theme--light" style="font-size: 70px; height: 70px; width: 70px;" aria-hidden="true"></i></transition-stub>
    </div>
  </div>
  <h1 data-v-c4ee72cd="" class="mt-10 text-h3 text-center">Today:</h1>
  <div data-v-c4ee72cd="" class="daily-agenda has-scroll mt-10 d-flex">
    <p data-v-c4ee72cd="" class="mx-auto text-h6">No trainings. Enjoy your day off!</p>
  </div>
  <p data-v-b4e148ca="" class="mt-10"><i data-v-b4e148ca="" class="mdi-github mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i><a data-v-b4e148ca="" class="text-grey-lighten-5 ml-2" target="_blank" noopener="" noreferrer="" href="https://www.github.com/ajuvonen/getfit">GitHub</a></p><button data-v-b4e148ca="" type="button" class="v-btn v-theme--light v-btn--density-default v-btn--size-x-small v-btn--variant-text home__photo-credit"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator="">Photo credits</span>
    <!---->
    <!---->
  </button>
  <!---->
  <!---->
</div>"
`;
